### DISCUSSION TITLE
casting address to struct pointer

### DISCUSSION THREAD

**Discussion by michaelsarr** [picocomputer/community#40] [Q&A]
Anyhow i did get a drawline functioning called from the 6502 and running in the pico  vga  

at the moment I have some code running on the 6502 

you create a struct 

struct msarrline
{
    uint16_t x0;
    uint16_t x1;
    uint16_t y0;
    uint16_t y1;
    uint8_t color;
};
Set the values for from x,y , to x,y and color then copy it to xram @ OxE000
 xreg(0xE00,16,1);// to fire it off
i put in a hook to call my function and pass it the address of the struct

vga side
```
uint16_t get_unit16(uint8_t * in)
{
    uint16_t x = (*(in+1) *  256) | *in;
    return x;
}
void vga_drawline(uint16_t struct_address)
{ 
uint16_t x0 = get_unit16(xram+struct_address);
uint16_t x1 = get_unit16(xram+struct_address+2); 
uint16_t y0 = get_unit16(xram+struct_address+4); 
uint16_t y1 = get_unit16(xram+struct_address+6); 
uint8_t color =  xram[struct_address+8];
/*
i could not seem to figure out how to cast the address coming in as an int to a pointer address to my structure type
everything I seemed to try came in with some other address , not something im familar with any suggestions would be apprciated
xram (pointer) + struct_address (int)  , comes out to the correct address 
 
i just had to pull the values from memory in the order they are defined on the client side 6502 structure

*/

int16_t dx =  abs (x1 - x0), sx = x0 < x1 ? 1 : -1;
int16_t dy = -abs (y1 - y0), sy = y0 < y1 ? 1 : -1; 
int16_t err = dx + dy, e2; /* error value e_xy */
 
  while(1)
  {  
    setPixel(x0,y0,color);
    if (x0 == x1 && y0 == y1) break;
    e2 = 2 * err;
    if (e2 >= dy) { err += dy; x0 += sx; } /* e_xy+e_x > 0 */
    if (e2 <= dx) { err += dx; y0 += sy; } /* e_xy+e_y < 0 */
  }
}// end drawline
 
void setPixel(int16_t x, int16_t y, uint8_t color) 
{

    uint16_t offset = (x / 2) + (y * (320 / 2));
    if (x%2)
     xram[offset] = (xram[offset] & 15) | (color << 4);
    else
      xram[offset] = (xram[offset] & 240) | color;
}
``` 
![IMG_6141](https://github.com/picocomputer/discussions/assets/73266612/e4ec40a4-61e0-4af2-aa3c-03a35da96cf9)

it is a whole lot faster, though

**Comment by rumbledethumps** [comment]
Cheater! This kind of acceleration wasn't common until the 90s. Did you measure the times difference?

**Reply by michaelsarr** [reply]
it draws lines from the center out to each pixel on the edge so (320x2 +240*2)  1120    lines drawn

running it all on the 6502 takes 32 seconds
passing the line draw off to the vga it takes .41 seconds
